integration-tests/tests/run-workflow.js

```diff
@@ -57,33 +59,49 @@ const launchWorkflowAndWaitForSuccess = async page => {
   ])
   // stopLoggingPageAjaxResponses()
 
+  // If this table doesn't exists, something is wrong. Fail test now.
+  await page.waitForXPath('//*[@role="table" and @aria-label="submission details"]', { visible: true })
+
+  // Workflow status list:
+  // Queued -> Submitted -> Launching -> Running -> Succeeded or Failed
   const start = Date.now()
+
+  // Long enough for the submission details to be Running or Failed.
+  // Workflows might sit in "Submitted" for a long time if there is a large backlog of workflows in the environment
+  const isRunning = await Promise.race([
+    findInGrid(page, 'Failed', { timeout: 5 * 60 * 1000 }).then(() => false),
+    findInGrid(page, 'Running', { timeout: 5 * 60 * 1000 }).then(() => true)
+  ])
+
+  // If status is not Running, fails test now
+  if (!isRunning) {
+    throw new Error('Workflow has failed')
+  }
+
+  // Wait until status is Succeeded or Failed
   await pRetry(async () => {
     try {
-      await findInGrid(page, 'Succeeded', { timeout: 65 * 1000 }) // long enough for the submission details to refresh
+      await Promise.race([
+        findInGrid(page, 'Succeeded', { timeout: 60 * 1000 }),
+        findInGrid(page, 'Failed', { timeout: 60 * 1000 })
```

Comment:
It looks like this changes the behavior of this test. Previously, the test expected the workflow to succeed. Now it looks like the test will pass if the workflow succeeds or fails. Is that intentional?

Reply:
no, you caught a bug. I added code to perform a final check after pRetry.
```
// pRetry will complete successfully if either Failed or Succeeded status if found.
  // We need to check status here to see if workflow has succeeded or not.
  try {
    await findInGrid(page, 'Failed', { timeout: 1000 })
    throw new Error('Workflow has failed')
  } catch (err) {
    // Succeeded
  }
 ```